### PR TITLE
cmake: turn libcommon into a shared library

### PR DESCRIPTION
**Pull request by tchaikov**
prior to this change, libcommon is a convenient library which gets
linked into librados, librbd and libcephfs and all ceph executables.
this incurs some problems:
 - double dose of libcommon in memory space and HDD: waste of memory
   and disk space.
 - if an application links to two libraries including libcommon at the
   same time. take librados and libcephfs as an example, they could
   interfere with each other by changing the other guy's status.
after this change, libcommon is tuned into a shared library and
renamed to libceph-common. it will be installed into $prefix/lib/ceph,
and packaged in librados2.

ceph.spec.in,debian/librados2.install: package libceph-common in
  librados2.
CMakeLists.txt:
  - do not link against libboost-* if not necessary.
  - s/common/ceph-common/g
  - install libceph-common into $prefix/lib/ceph
  - set rpath to $prefix/lib/ceph
  - link against ceph-common if an executable needs access to non public
    symbols in ceph.

Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -535,16 +538,38 @@ else()
 endif()
 
 if(ENABLE_SHARED)
-  list(APPEND libcommon_files
-    $<TARGET_OBJECTS:global_common_objs>)
 endif(ENABLE_SHARED)
```
Comment (rjfd) [inline src/CMakeLists.txt]: nit: why keep this empty conditional block?

File: CMakeLists.txt [hunk 2]
```diff
@@ -518,15 +518,6 @@ find_package(Boost 1.61 COMPONENTS ${BOOST_COMPONENTS} REQUIRED)
 include_directories(SYSTEM ${Boost_INCLUDE_DIRS})
 include_directories(SYSTEM ${PROJECT_BINARY_DIR}/include)
 
-if (WITH_SYSTEM_BOOST)
```
Comment (rjfd) [inline CMakeLists.txt]: Was this necessary for turning libcommon into a shared library, or is just a cleanup? If it's just to clean up, maybe we should separate it from this PR.

Comment (tchaikov) [inline CMakeLists.txt]: @rjfd from a first glance, yeah, it is indeed a cleanup. but it was a side effect of turning libcommon into a share lib. i forgot exactly why.

Comment (tchaikov) [inline CMakeLists.txt]: oh, i moved the linkage to ceph-common.
```diff
+if(NOT WITH_SYSTEM_BOOST)
+  target_link_libraries(ceph-common ${ZLIB_LIBRARIES})
+endif()
```

**Comment by tchaikov** [comment]
changelog

- remove empty if block in `CMakeLists.txt`.

**Comment by liewegas** [comment]
:+1: on this change!  This should avoid some longstanding headaches.

**Comment by rjfd** [comment]
@tchaikov will you run a set of suites in teuthology to check if nothing stops working?

**Comment by tchaikov** [comment]
@rjfd sure, i am on it.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-01-10_14:07:09-rados-wip-kefu-testing---basic-mira/

**Comment by smithfarm** [comment]
When I try to build this in the OBS, I get:

```
[ 3178s] -- Install configuration: "RelWithDebInfo"
[ 3178s] -- Installing: /home/abuild/rpmbuild/BUILDROOT/ceph-11.1.0+git.1484223486.fa61e0b-1.1.x86_64/usr/lib64/ceph/libceph-common.so
[ 3179s] CMake Error at src/cmake_install.cmake:53 (file):
[ 3179s]   file RPATH_CHANGE could not write new RPATH:
[ 3179s] 
[ 3179s]     /usr/lib64/ceph
[ 3179s] 
[ 3179s]   to the file:
[ 3179s] 
[ 3179s]     /home/abuild/rpmbuild/BUILDROOT/ceph-11.1.0+git.1484223486.fa61e0b-1.1.x86_64/usr/lib64/ceph/libceph-common.so
[ 3179s] 
[ 3179s]   No valid ELF RPATH or RUNPATH entry exists in the file;
[ 3179s] Call Stack (most recent call first):
[ 3179s]   cmake_install.cmake:37 (include)
```

**Comment by tchaikov** [comment]
@smithfarm could you try to build it with `-DCMAKE_BUILD_WITH_INSTALL_RPATH=ON`? seems the rpath of `libceph-common.so` was removed at build-time, so cmake failed to update its rpath at install-time. but the downside is, the built executable might not be able to run in the build directory.

**Comment by tchaikov** [comment]
see also http://lists.llvm.org/pipermail/llvm-dev/2014-February/070536.html

**Pull request merged**